threads_win: fix build error with VS2010

httpstorm opened the pull request:
As `snprintf` is not available everywhere, use `BIO_snprintf` instead.
Fixes:
```
        IF EXIST test\quic_multistream_test.exe.manifest DEL /F /Q test\quic_multistream_test.exe.manifest
        "link" /nologo /debug setargv.obj /subsystem:console /opt:ref  /nologo /debug @V:\_tmp\nm4.tmp
quic_multistream_test-bin-quic_multistream_test.obj : error LNK2019: unresolved external symbol _snprintf referenced in function _helper_init
test\quic_multistream_test.exe : fatal error LNK1120: 1 unresolved externals
NMAKE : fatal error U1077: '"E:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\BIN\link.EXE"' : return code '0x460'
```

VC 2010 or earlier compilers do not support `static inline`. To work around this problem, we can use the `ossl_inline` macro.

Fixes:
```
crypto\threads_win.c(171) : error C2054: expected '(' to follow 'inline'
crypto\threads_win.c(172) : error C2085: 'get_hold_current_qp' : not in formal parameter list
crypto\threads_win.c(172) : error C2143: syntax error : missing ';' before '{'
crypto\threads_win.c(228) : warning C4013: 'get_hold_current_qp' undefined; assuming extern returning int
crypto\threads_win.c(228) : warning C4047: '=' : 'rcu_qp *' differs in levels of indirection from 'int'
```

`InterlockedAnd64` and `InterlockedAdd64` are not available on VS2010 x86. We already have implemented replacements for other functions, such as `InterlockedOr64`. Apply the same approach to fix the errors. A `CRYPTO_RWLOCK rw_lock` is added to `rcu_lock_st`.

Note that the return values of the replacement atomic functions for VS2010 x86 are not checked, because I did not know how to respond to these errors in each location where they are used. This might happen if the lock is not initialised yet. Though the lock is initialised in the beginning when each object is constructed, so this should not occur. Please let me know if any changes are needed.

Fixes:
```
libcrypto.lib(libcrypto-lib-threads_win.obj) : error LNK2019: unresolved external symbol _InterlockedAdd64 referenced in function _get_hold_current_qp
libcrypto.lib(libcrypto-lib-threads_win.obj) : error LNK2019: unresolved external symbol _InterlockedOr referenced in function _get_hold_current_qp
libcrypto.lib(libcrypto-lib-threads_win.obj) : error LNK2019: unresolved external symbol _InterlockedAnd64 referenced in function _update_qp
libcrypto.lib(libcrypto-lib-threads_win.obj) : error LNK2019: unresolved external symbol _InterlockedOr64 referenced in function _ossl_synchronize_rcu
```

Build and run tested: Windows 2003 x64, Windows 2003 x86, Visual Studio 2010.
@t8m @levitte @paulidale @nhorman @mattcaswell @hlandau 

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

httpstorm commented:
I submitted signed CLA over e-mail.

nhorman commented on crypto/threads_win.c:
```diff
@@ -143,6 +144,7 @@ CRYPTO_RCU_LOCK *ossl_rcu_lock_new(int num_writers, OSSL_LIB_CTX *ctx)
         return NULL;
 
     new->ctx = ctx;
+    new->rw_lock = CRYPTO_THREAD_lock_new();
```
you need to check this for null in the if conditional below, or you'll have a leak

httpstorm replied:
I added the check.

nhorman commented on crypto/threads_win.c:
```diff
@@ -569,6 +576,44 @@ int CRYPTO_atomic_add(int *val, int amount, int *ret, CRYPTO_RWLOCK *lock)
     return 1;
 }
 
+int CRYPTO_atomic_add64(uint64_t *val, uint64_t op, uint64_t *ret,
```
you need to run make update to get this in the exported ordinals list, its why CI is failing here

httpstorm replied:
Thanks! I was wondering what this error means.
That ancient piece of history Windows 2003 is using `nmake` to compile, so I ran the command on macOS instead. I see the two new functions were added to `util/libcrypto.num`. I hope that's good enough.

nhorman replied:
Not quite.  The CI test for symbol presence is failing, need to look at why.

also, if you're going to add new exported library symbols, the docs check is going to fail if you don't document them in the appropriate pod file

httpstorm replied:
`doc/man3/CRYPTO_THREAD_run_once.pod` contains all related function, so I added the description there.

httpstorm replied:
Building on Windows works only with `nmake`. I do have `GNU Make 4.4.1`, but both fail to perform the `update` target. So my options are macOS 14.4.1 Intel, and Ubuntu 22.04.4 LTS.

```
nmake update
Microsoft (R) Program Maintenance Utility Version 10.00.40219.01
Copyright (C) Microsoft Corporation.  All rights reserved.

NMAKE : fatal error U1073: don't know how to make 'update'
Stop.

make update
makefile:78: *** missing separator.  Stop.
```

I also noticed that apart from `threads_win.c`, the atomic functions are also defined in `threads_pthread.c` and `threads_none.c`. Do I need to implement them there as well?

httpstorm replied:
Or maybe I need to add them to `./test/threadstest.c`?

nhorman replied:
you do need to add the definitions to threads_pthread.c and threads_none.c, as only one of the three (win/pthreads/none) gets compiled for the build and you need a definition for the exported symbols

As for why make update isn't working.  I expect its got something to do with the version of nmake on vs2010.  Its worked for me on VS2022 community edition.  note VS2010 isn't an officially [supported platform](https://www.openssl.org/policies/general-supplemental/platforms.html)

I'd focus on getting the implementations defined on other platforms first, then worry about being able to make the update script.

httpstorm replied:
Added definitions to `threads_pthread.c` and `threads_none.c`. Added tests to `threadstest.c`.

nhorman commented:
whats the output of:
```
make TESTS=test_threads V=1 test
```
with this change?  I'd like to make sure this isn't going to create a performance impact to the rcu locks on windows

nhorman commented on crypto/threads_win.c:
```diff
@@ -154,7 +156,8 @@ CRYPTO_RCU_LOCK *ossl_rcu_lock_new(int num_writers, OSSL_LIB_CTX *ctx)
         || new->alloc_lock == NULL
-        || new->prior_lock == NULL) {
+        || new->prior_lock == NULL
+        || new->rw_lock == NULL) {
```
you also need to free rw_lock here.  e.g if you successfully allocate rw_lock but prior_lock returns null, you will still get  a leak.  Its ok if you call an objects free function with a NULL pointer, they all check for that

httpstorm replied:
I added a free call in case of an error.

httpstorm commented:
> whats the output of:
> 
> ```
> make TESTS=test_threads V=1 test
> ```

- [x] CI tests pass
- [x] [test_threads.2024-05-04.01.ubuntu-22.04.4.LTS.txt](https://httpstorm.com/share/.ssl/test_threads.2024-05-04.01.ubuntu-22.04.4.LTS.txt)
- [x] [test_threads.2024-05-04.02.macOS-14.4.1.x64.txt](https://httpstorm.com/share/.ssl/test_threads.2024-05-04.02.macOS-14.4.1.x64.txt)
- [x] [test_threads.2024-05-04.03.FreeBSD-13.2-p11.txt](https://httpstorm.com/share/.ssl/test_threads.2024-05-04.03.FreeBSD-13.2-p11.txt)
- [x] [test_threads.2024-05-04.04.Windows-11-21H2.VS2019.txt](https://httpstorm.com/share/.ssl/test_threads.2024-05-04.04.Windows-11-21H2.VS2019.txt)

- [ ] [test_threads.2024-05-04.05.Windows-2003-x64.VS2010.txt](https://httpstorm.com/share/.ssl/test_threads.2024-05-04.05.Windows-2003-x64.VS2010.txt) fails to compile tests:
``` batch
        IF EXIST test\quic_multistream_test.exe.manifest DEL /F /Q test\quic_multistream_test.exe.manifest
        "link" /nologo /debug setargv.obj /subsystem:console /opt:ref  /nologo /debug @V:\_tmp\nm4.tmp
quic_multistream_test-bin-quic_multistream_test.obj : error LNK2019: unresolved external symbol _snprintf referenced in function _helper_init
test\quic_multistream_test.exe : fatal error LNK1120: 1 unresolved externals
NMAKE : fatal error U1077: '"E:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\BIN\link.EXE"' : return code '0x460'
```

> with this change? I'd like to make sure this isn't going to create a performance impact to the rcu locks on windows

I agree, performance is an important factor. One approach would be to use macros, so that only VS2010 x86 target is affected. I didn't do it this way, because I tried to follow existing patterns. We can also apply this to other atomic functions (that are already replaced) if you find it necessary.

Note: macOS test ran on MacBook Pro 2019 Intel, the rest of the tests ran on Intel i7 930.
I will review the changes once again after a good sleep.

nhorman commented:
Something isn't right, your tests on windows 11 show that RCU lock performance is about 1/2 of corresponding rw lock performance.  They should.be about equal for the light contention case and RCU should do much better in the heavy contention case.

httpstorm commented:
@nhorman 

> Something isn't right, your tests on windows 11 show that RCU lock performance is about 1/2 of corresponding rw lock performance. They should.be about equal for the light contention case and RCU should do much better in the heavy contention case.

Any idea why? I see the same machine on Linux matches the behaviour you mentioned.
The Windows 11 tests from last night use configuration options I copied from the Windows 2022 runner.
```
perl Configure --banner=Configured enable-demos no-makedepend no-shared no-fips enable-md2 enable-rc5 enable-ssl3 enable-ssl3-method enable-weak-ssl-ciphers enable-trace enable-crypto-mdebug VC-WIN64A-masm
```

The computer is from 2010 i7 930 4 cores+HT. I ran that test again to compare performance before and after the patches on the same machine, while idle. RCU performance is again half of RW.
`perl Configure VC-WIN64A no-shared`
- [test_threads.2024-05-04.07.Windows-11-21H2.VS2019.before.txt](https://httpstorm.com/share/.ssl/test_threads.2024-05-04.07.Windows-11-21H2.VS2019.before.txt)
- [test_threads.2024-05-04.08.Windows-11-21H2.VS2019-after.txt](https://httpstorm.com/share/.ssl/test_threads.2024-05-04.08.Windows-11-21H2.VS2019-after.txt)
``` diff
# torture_rw_low
- performed 47544198 reads and 6 writes over 2 read and 2 write threads in 4.056000e+00 seconds
+ performed 48842838 reads and 6 writes over 2 read and 2 write threads in 4.048000e+00 seconds
- Average read time 8.531009e-08/read
+ Average read time 8.287807e-08/read
- Averate write time 6.760000e-01/write
+ Averate write time 6.746667e-01/write

# torture_rw_high
- performed 41121498 reads and 654482 writes over 2 read and 2 write threads in 3.589000e+00 seconds
+ performed 43130985 reads and 795234 writes over 2 read and 2 write threads in 3.920000e+00 seconds
- Average read time 8.727795e-08/read
+ Average read time 9.088594e-08/read
- Averate write time 5.483726e-06/write
+ Averate write time 4.929367e-06/write

# torture_rcu_low
- performed 24620885 reads and 6 writes over 2 read and 2 write threads in 4.044000e+00 seconds
+ performed 25965189 reads and 6 writes over 2 read and 2 write threads in 4.055000e+00 seconds
- Average read time 1.642508e-07/read
+ Average read time 1.561706e-07/read
- Average write time 6.740000e-01/write
+ Average write time 6.758333e-01/write

# torture_rcu_high
- performed 21784861 reads and 1760318 writes over 2 read and 2 write threads in 3.956000e+00 seconds
+ performed 25114008 reads and 1759286 writes over 2 read and 2 write threads in 3.945000e+00 seconds
- Average read time 1.815940e-07/read
+ Average read time 1.570836e-07/read
- Average write time 2.247321e-06/write
+ Average write time 2.242387e-06/write

# Linux
# torture_rw_low
performed 29389229 reads and 6 writes over 2 read and 2 write threads in 4.001009e+00 seconds
Average read time 1.361386e-07/read
Averate write time 6.668348e-01/write

# torture_rw_high
performed 8561309 reads and 336462 writes over 2 read and 2 write threads in 3.367428e+00 seconds
Average read time 3.933310e-07/read
Averate write time 1.000835e-05/write

# torture_rcu_low
performed 26298285 reads and 6 writes over 2 read and 2 write threads in 4.000877e+00 seconds
Average read time 1.521345e-07/read
Average write time 6.668128e-01/write

# torture_rcu_high
performed 34343847 reads and 2291988 writes over 2 read and 2 write threads in 3.998745e+00 seconds
Average read time 1.164326e-07/read
Average write time 1.744662e-06/write
```

Good news: Windows 2003 + VS2010 pass the `test_threads`.
- [x] [test_threads.2024-05-04.06.Windows-2003-x64.VS2010.txt](https://httpstorm.com/share/.ssl/test_threads.2024-05-04.06.Windows-2003-x64.VS2010.txt)
- [ ] [all.2024-05-04.05.Windows-2003-x64.VS2010.txt](https://httpstorm.com/share/.ssl/all.2024-05-04.05.Windows-2003-x64.VS2010.txt) All tests: 1 fails: `04-test_bio_dgram.t`

Note: the IPv6 network stack was not installed. If I install it, that test passes, but later two other fail
- [ ] [all.2024-05-04.06.Windows-2003-x64.VS2010-ipv6.txt](https://httpstorm.com/share/.ssl/all.2024-05-04.06.Windows-2003-x64.VS2010-ipv6.txt)

I ran all tests on other platforms, here are the results. Note that on Windows x64 and x86 tests were ran at the same time.
- [x] [all.2024-05-04.01.ubuntu-22.04.4.LTS.txt](https://httpstorm.com/share/.ssl/all.2024-05-04.01.ubuntu-22.04.4.LTS.txt)
- [x] [all.2024-05-04.02.macOS-14.4.1.x64.txt](https://httpstorm.com/share/.ssl/all.2024-05-04.02.macOS-14.4.1.x64.txt)
- [x] [all.2024-05-04.03.FreeBSD-13.2-p11.txt](https://httpstorm.com/share/.ssl/all.2024-05-04.03.FreeBSD-13.2-p11.txt)
- [x] [all.2024-05-04.04.Windows-11-21H2.VS2019.txt](https://httpstorm.com/share/.ssl/all.2024-05-04.04.Windows-11-21H2.VS2019.txt)

quic_multistream_test: fix undefined symbol snprintf with VS2010
Fixes building of tests during OpenSSL compile.
No regression on other platforms. The tests above are compiled with this patch.
``` diff
diff --git a/test/quic_multistream_test.c b/test/quic_multistream_test.c
index e01e8ae85a..8437d879f9 100644
--- a/test/quic_multistream_test.c
+++ b/test/quic_multistream_test.c
@@ -23,6 +23,12 @@
 #endif
 #include "internal/numbers.h"  /* UINT64_C */
 
+#if (defined(_MSC_VER) && _MSC_VER <= 1600)
+# define SNPRINTF_FN sprintf_s
+#else
+# define SNPRINTF_FN snprintf
+#endif
+
 static const char *certfile, *keyfile;
 
 #if defined(OPENSSL_THREADS)
@@ -784,7 +790,7 @@ static int helper_init(struct helper *h, const char *script_name,
         goto err;
 
     /* Set title for qlog purposes. */
-    snprintf(title, sizeof(title), "quic_multistream_test: %s", script_name);
+    SNPRINTF_FN(title, sizeof(title), "quic_multistream_test: %s", script_name);
     if (!TEST_true(ossl_quic_set_diag_title(h->c_ctx, title)))
         goto err;
 
@@ -5827,7 +5833,7 @@ static int test_script(int idx)
     }
 #endif
 
-    snprintf(script_name, sizeof(script_name), "script %d", script_idx + 1);
+    SNPRINTF_FN(script_name, sizeof(script_name), "script %d", script_idx + 1);
 
     TEST_info("Running script %d (order=%d, blocking=%d)", script_idx + 1,
               free_order, blocking);
@@ -5912,7 +5918,7 @@ static ossl_unused int test_dyn_frame_types(int idx)
             s[i].arg2 = forbidden_frame_types[idx].expected_err;
         }
 
-    snprintf(script_name, sizeof(script_name),
+    SNPRINTF_FN(script_name, sizeof(script_name),
              "dyn script %d", idx);
 
     return run_script(dyn_frame_types_script, script_name, 0, 0);
-- 
2.44.0
```

I noticed that existing atomic functions in `threads_pthread.c` check for `if (ret != NULL)` only in one of the precompiled cases. Is this intentional or should it be corrected? I followed the same implementation in the newly added functions. Such a check is not performed in `threads_none.c` and `threads_win.c`, which is inconsistent. The documentation `CRYPTO_THREAD_run_once.pod` does not mention that `ret` is checked for NULL, only `lock`.

Please let me know if I need to change anything else. What about the `snprintf` patch?

viruscamp commented:
@httpstorm  please check  #24109 , I have built openssl-3.3.0 with that patch in vs2010 x86 winxp-sp3-x86.  
Another change is replacing `snprintf` with `BIO_snprintf`.  

The tests results:  
- Without ipv6 enabled, Failed at 04-test_bio_dgram.t , just same as #19142  
- With ipv6 enabled,  04-test_bio_dgram.t was passed, but 10 more tests failed all about IPV6

httpstorm commented:
@viruscamp 
Thank you! Indeed an assembly implementation would be way faster and we won't have to make so many changes.
I am traveling for a few days. I will test the assembly patch and take care of `snprintf`, once I'm back.

t8m commented on crypto/threads_win.c:
```diff
@@ -168,17 +184,23 @@ void ossl_rcu_lock_free(CRYPTO_RCU_LOCK *lock)
 }
 
-static inline struct rcu_qp *get_hold_current_qp(CRYPTO_RCU_LOCK *lock)
+static INLINE_VS struct rcu_qp *get_hold_current_qp(CRYPTO_RCU_LOCK *lock)
```
We have `ossl_inline` please use that.

httpstorm replied:
Thanks! I changed it. Test build successful on Windows 2003, VS2010.

httpstorm commented on crypto/threads_win.c:
```diff
@@ -27,6 +27,17 @@
+#else
+# define INLINE_VS inline
+#endif
+
```
Dropped. Now using `ossl_inline`.

tom-cosgrove-arm commented on crypto/threads_win.c:
```diff
@@ -103,6 +103,7 @@ struct rcu_lock_st {
     CRYPTO_CONDVAR *alloc_signal;
     CRYPTO_MUTEX *prior_lock;
     CRYPTO_CONDVAR *prior_signal;
+    CRYPTO_RWLOCK * rw_lock;
```
There should really not be a space after the `*` here

httpstorm replied:
Fixed

httpstorm commented:
Updated:
- use `BIO_snprintf` in `quic_multistream_test`
- use `ossl_inline` in `threads_win.c`

Tests: https://httpstorm.com/share/.ssl/

Now about the ASM lockless implementations of atomic functions: I can either
1. replace the last patch with them
2. integrate them in the replacement function, only those I added
3. integrate them in the replacement function, all
4. do nothing

tom-cosgrove-arm commented:
I'm becoming involved with this at a late stage, so don't want to be very demanding, but it seems that this PR does three different things:

1) use `BIO_snprintf()` consistently, since not all platforms have `snprintf`;
2) use `ossl_inline` consistently, since not all compilers support `inline`;
3) deal with `InterlockedAnd64()` and `InterlockedAdd64()` not being available on older versions of Visual Studio.

While all 3 are needed to get OpenSSL building with older versions of Visual Studio, they feel like quite separate things.

I personally think that 3 separate PRs would mean (1) and (2) could be merged very quickly, while we [consider the best approach for the missing lockless functions](https://github.com/openssl/openssl/pull/24326#issuecomment-2100279348).

If other people would rather this stays as one PR that's fair enough; splitting this might get things merged sooner

t8m commented:
I would also support splitting this as 1 and 2 would be eligible for cherry pick to stable branches but 3 does not really look like so.

httpstorm commented:
@tom-cosgrove-arm @t8m @nhorman @viruscamp
- part 1/3 https://github.com/openssl/openssl/pull/24369
- part 2/3 https://github.com/openssl/openssl/pull/24370
- part 3/3 I can rebase this PR or create a new one after part 2 is accepted. 1 and 2 are independent.

I took time to analyse the ASM code from https://github.com/openssl/openssl/issues/24109 . It is better optimised than what Visual Studio 2019 would generate for a x86 release (I stepped it in the debugger). The code reads `_Ptr`, adds `_Val` then performs atomic compare exchange:
- if the original value at `_Ptr` did not change, the new value is written and the function completes
- if the original value at `_Ptr` did change, it starts over

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago
